hostdiscovery/hostdiscovery.go

```diff
@@ -0,0 +1,160 @@
+package hostdiscovery
+
+import (
+	"errors"
+	"net"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+
+	mapset "github.com/deckarep/golang-set"
+	"github.com/docker/libnetwork/config"
+	"github.com/docker/swarm/discovery"
+	// Annonymous import will be removed after we upgrade to latest swarm
+	_ "github.com/docker/swarm/discovery/file"
+	// Annonymous import will be removed after we upgrade to latest swarm
+	_ "github.com/docker/swarm/discovery/kv"
+	// Annonymous import will be removed after we upgrade to latest swarm
+	_ "github.com/docker/swarm/discovery/nodes"
+	// Annonymous import will be removed after we upgrade to latest swarm
+	_ "github.com/docker/swarm/discovery/token"
+)
+
+const defaultHeartbeat = 10
+
+// JoinCallback provides a callback event for new node joining the cluster
+type JoinCallback func(entries []net.IP)
+
+// LeaveCallback provides a callback event for node leaving the cluster
+type LeaveCallback func(entries []net.IP)
+
+// HostDiscovery primary interface
+type HostDiscovery interface {
+	StartDiscovery(*config.ClusterCfg, JoinCallback, LeaveCallback) error
+	StopDiscovery() error
+	Fetch() ([]net.IP, error)
+}
+
+type hostDiscovery struct {
+	discovery discovery.Discovery
+	nodes     []net.IP
+	stopChan  chan struct{}
+	sync.Mutex
+}
+
+// NewHostDiscovery function creates a host discovery object
+func NewHostDiscovery() HostDiscovery {
+	return &hostDiscovery{nodes: []net.IP{}, stopChan: make(chan struct{})}
+}
+
+func (h *hostDiscovery) StartDiscovery(cfg *config.ClusterCfg, joinCallback JoinCallback, leaveCallback LeaveCallback) error {
+	hb := cfg.Heartbeat
+	if hb == 0 {
+		hb = defaultHeartbeat
+	}
+	d, err := discovery.New(cfg.Discovery, hb)
+	if err != nil {
+		return err
+	}
+
+	go d.Watch(func(entries []*discovery.Entry) {
+		h.processCallback(entries, joinCallback, leaveCallback)
+	})
+
+	if ip := net.ParseIP(cfg.Address); ip == nil {
```

Comment:
You should probably do these checks and the below `d.Register` before starting the goroutine because if you are failing here we have a goroutine leak in our hands